Document UrlTokenReplaceEvent for URL token replacement (#517)
mautic/developer-documentation-new: docs/plugin_extensions/landing_pages.rst

Reviewer: ```suggestion
Landing Pages
```

Reviewer: ```suggestion
There are two way to extend Landing Pages:

* Landing Page tokens used to insert Dynamic Content into a Landing Page
* A/B test winning criteria
```

.. vale on

> There are two way to extend Landing Pages:

Reviewer: Small grammar fix here.

```suggestion
There are two ways to extend Landing Pages:
```

.. vale off

Landing page tokens

> *******************

Reviewer: Landing Page is Mautic's feature. It should be capitalized.

```suggestion
.. vale off

Landing Page tokens
*******************

.. vale on
```

.. vale off

Page A/B test winner criteria
*****************************

> Custom Landing Page A/B test winner criteria get handled exactly the same as :ref:`Email A/B testing`, except the callback function receives the following arguments:

Reviewer: @promptless-for-oss use inline code for 'callback'.

```suggestion
Custom Landing Page A/B test winner criteria get handled exactly the same as :ref:`Email A/B testing`, except the ``callback`` function receives the following arguments:
```

.. vale off

Page A/B test winner criteria
*****************************

> Custom Landing Page A/B test winner criteria get handled exactly the same as :ref:`Email A/B test winner criteria<components/emails:Email tokens and A/B testing>` with the only differences being that the ``callback`` function gets passed ``Mautic\PageBundle\Entity\Page $page`` and ``Mautic\PageBundle\Entity\Page $parent`` instead.

Reviewer: The ref link here is incorrect, and it's a bit hard to read. We need to fix this:

- In the `plugin_extensions/emails.rst`, right above the "Email A/B testing" title, add custom target: `.. _Email A/B testing:`
- Replace the incorrect link with the target link.
- Maybe we can use bullet points for the arguments for easier read.

```suggestion
Custom Landing Page A/B test winner criteria get handled exactly the same as :ref:`Email A/B testing`, except the callback function receives the following arguments:

* ``Mautic\PageBundle\Entity\Page $page``
* ``Mautic\PageBundle\Entity\Page $parent``
```

            $event->setContent($content);
        }
    }

.. vale off

URL Token Replacement

> *********************

Reviewer: Use sentence case for titles.

```suggestion
URL token replacement
*********************
```

@promptless-for-oss please address the suggestions.

.. vale on

> When a Contact clicks a tracked link, Mautic's redirect handler can replace Contact field tokens in the target URL before redirecting. This allows personalized destination URLs based on Contact data.

Author: Added documentation for `UrlTokenReplaceEvent` based on PR #16199. The PR introduces this new event in `PublicController::redirectAction()` and moves Contact field token replacement to a `PageSubscriber` event subscriber. The event provides `getContent()`, `setContent()`, `getLead()`, and `getEmailId()` methods as shown in `UrlTokenReplaceEvent.php`.

Source: https://github.com/mautic/mautic/pull/16199

The ``Mautic\PageBundle\Event\UrlTokenReplaceEvent`` event dispatches during the redirect process. Use it to customize or extend URL token replacement logic. By default, Mautic's built-in subscriber handles standard Contact field tokens like ``{contactfield=email}`` or ``{contactfield=firstname}``.

Use cases for this event include:

- Adding custom token types beyond Contact fields
- Transforming token values before URL insertion
- Implementing conditional token replacement based on Email context

> - Logging or auditing token replacements

Reviewer: ```suggestion
* Adding custom token types beyond Contact fields
* Transforming token values before URL insertion
* Implementing conditional token replacement based on Email context
* Logging or auditing token replacements
```

The event provides:

- ``getContent()`` - Returns the current URL string
- ``setContent(string $url)`` - Sets the modified URL
- ``getLead()`` - Returns the ``Mautic\LeadBundle\Entity\Lead`` Contact entity

> - ``getEmailId()`` - Returns the Email ID if the redirect originated from an Email click, or ``null`` otherwise

Reviewer: ```suggestion
* ``getContent()`` - Returns the current URL string
* ``setContent(string $url)`` - Sets the modified URL
* ``getLead()`` - Returns the ``Mautic\LeadBundle\Entity\Lead`` Contact entity
* ``getEmailId()`` - Returns the Email ID if the redirect originated from an Email click, or ``null`` otherwise
```

    {
        public static function getSubscribedEvents(): array
        {
            return [
                UrlTokenReplaceEvent::class => ['onUrlTokenReplace', 0],
            ];
        }

        public function onUrlTokenReplace(UrlTokenReplaceEvent $event): void
        {
            $url = $event->getContent();

            // Check if URL contains tokens using the public regex
            if (!preg_match(TokenHelper::REGEX, $url)) {
                return;
            }

            $contact = $event->getLead();
            $emailId = $event->getEmailId();

            // Implement custom token replacement logic
            // For example, replace a custom token with Contact data
            if (str_contains($url, '{custom_token}')) {
                $customValue = $this->getCustomValue($contact, $emailId);
                $url = str_replace('{custom_token}', urlencode($customValue), $url);
                $event->setContent($url);
            }
        }

        private function getCustomValue($contact, ?int $emailId): string
        {
            // Your custom logic here
            return 'custom_value';
        }
    }

.. note::

>     The ``TokenHelper::REGEX`` constant provides the regular expression pattern for matching Contact field tokens. Use this to check whether a URL contains tokens before performing replacements.

Reviewer: ```suggestion
.. note::

   The ``TokenHelper::REGEX`` constant provides the regular expression pattern for matching Contact field tokens. Use this to check whether a URL contains tokens before performing replacements.
```
